File: services/clinical-service/src/domain/entities/ClinicalNote.ts
import { v4 as uuidv4 } from 'uuid';

export enum NoteType {
    PROGRESS = 'PROGRESS',
    ADMISSION = 'ADMISSION',
    DISCHARGE = 'DISCHARGE',
    CONSULTATION = 'CONSULTATION',
    PROCEDURE = 'PROCEDURE',
}

export interface ClinicalNoteProps {
    id?: string;
    encounterId: string;
    content: string;
    authorId: string;
    noteType: NoteType;
    createdAt?: Date;
}

export class ClinicalNote {
    readonly id: string;
    readonly encounterId: string;
    readonly content: string;
    readonly authorId: string;
    readonly noteType: NoteType;
    readonly createdAt: Date;

    private constructor(props: Required<ClinicalNoteProps>) {
        this.id = props.id;
        this.encounterId = props.encounterId;
        this.content = props.content;
        this.authorId = props.authorId;
        this.noteType = props.noteType;
        this.createdAt = props.createdAt;
    }

    static create(props: ClinicalNoteProps): ClinicalNote {
        if (!props.content || props.content.trim().length === 0) {
            throw new Error('Clinical note content cannot be empty');
        }

        return new ClinicalNote({
            id: props.id ?? uuidv4(),
            encounterId: props.encounterId,
            content: props.content,
            authorId: props.authorId,
            noteType: props.noteType,
            createdAt: props.createdAt ?? new Date(),
        });
    }
}

File: services/clinical-service/src/main/java/com/medtrust/clinical/domain/model/ClinicalNote.java
package com.medtrust.clinical.domain.model;

import java.time.Instant;
import java.util.UUID;

public class ClinicalNote {

    private final String id;
    private final String encounterId;
    private final String content;
    private final String authorId;
    private final NoteType noteType;
    private final Instant createdAt;

    private ClinicalNote(String id, String encounterId, String content,
            String authorId, NoteType noteType, Instant createdAt) {
        this.id = id;
        this.encounterId = encounterId;
        this.content = content;
        this.authorId = authorId;
        this.noteType = noteType;
        this.createdAt = createdAt;
    }

    public static ClinicalNote create(String encounterId, String content,
            String authorId, NoteType noteType) {
        if (content == null || content.isBlank()) {
            throw new IllegalArgumentException("Clinical note content cannot be empty");
        }
        return new ClinicalNote(
                UUID.randomUUID().toString(),
                encounterId,
                content,
                authorId,
                noteType,
                Instant.now());
    }

    public static ClinicalNote reconstitute(String id, String encounterId, String content,
            String authorId, NoteType noteType, Instant createdAt) {
        return new ClinicalNote(id, encounterId, content, authorId, noteType, createdAt);
    }

    public String getId() {
        return id;
    }

    public String getEncounterId() {
        return encounterId;
    }

    public String getContent() {
        return content;
    }

    public String getAuthorId() {
        return authorId;
    }

    public NoteType getNoteType() {
        return noteType;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }
}
